src/benchmarklib/benchmark_runner.cpp

```diff
@@ -0,0 +1,449 @@
+    if (query_plans_iter == _query_plans.end()) {
+      Assert(pipeline.get_query_plans().size() == 1, "Expected exactly one SQLQueryPlan");
+      QueryPlans plans{pipeline.get_optimized_logical_plans(), pipeline.get_query_plans()};
+      _query_plans.emplace(name, plans);
+    }
+  }
+}
+
+void BenchmarkRunner::_create_report(std::ostream& stream) const {
+  nlohmann::json benchmarks;
+
+  for (const auto& named_query : _queries) {
+    const auto& name = named_query.first;
+    const auto& query_result = _query_results_by_query_name.at(name);
+
+    const auto duration_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(query_result.duration).count();
+    const auto duration_seconds = static_cast<float>(duration_ns) / 1'000'000'000;
+    const auto items_per_second = static_cast<float>(query_result.num_iterations) / duration_seconds;
+    const auto time_per_query = duration_ns / query_result.num_iterations;
+
+    nlohmann::json benchmark{
+        {"name", name},
+        {"iterations", query_result.num_iterations},
+        {"real_time", time_per_query},
+        {"cpu_time", time_per_query},
+        {"items_per_second", items_per_second},
+        {"time_unit", "ns"},
+    };
+
+    benchmarks.push_back(benchmark);
+  }
+
+  nlohmann::json report{{"context", _context}, {"benchmarks", benchmarks}};
+
+  stream << std::setw(2) << report << std::endl;
+}
+
+BenchmarkRunner BenchmarkRunner::create_tpch(BenchmarkConfig config, const std::vector<QueryID>& query_ids,
+                                             const float scale_factor) {
+  NamedQueries queries;
+  queries.reserve(query_ids.size());
+
+  for (const auto query_id : query_ids) {
+    queries.emplace_back("TPC-H " + std::to_string(query_id), tpch_queries.at(query_id));
+  }
+
+  config.out << "- Generating TPCH Tables with scale_factor=" << scale_factor << "..." << std::endl;
+
+  ColumnEncodingSpec encoding_spec{config.encoding_type};
+  const auto tables = opossum::TpchDbGenerator(scale_factor, config.chunk_size).generate();
+
+  for (auto& table : tables) {
+    if (config.encoding_type != EncodingType::Unencoded) {
+      ChunkEncoder::encode_all_chunks(table.second, encoding_spec);
+    }
+
+    StorageManager::get().add_table(tpch_table_names.at(table.first), table.second);
+  }
+  config.out << "- Done." << std::endl;
+
+  auto context = _create_context(config);
+  // Add TPCH-specific information
+  context.emplace("scale_factor", scale_factor);
+
+  return BenchmarkRunner(std::move(config), queries, context);
+}
+
+BenchmarkRunner BenchmarkRunner::create(BenchmarkConfig config, const std::string& table_path,
+                                        const std::string& query_path) {
+  const auto tables = _parse_table_path(table_path);
+  if (tables.empty()) {
+    throw std::runtime_error("No tables found in '" + table_path + "'");
+  }
+
+  CsvMeta csv_meta{};
+  csv_meta.chunk_size = config.chunk_size;
```

Comment:
Have you actually tested this? When setting CsvMeta, you need to specify the column as well.
Tbh, that does not make too much sense for me. If no columns are set, the CsvImport should take the first line as the header.

Reply:
the main problem is that the config can contain different chunk_sizes then specified in the benchmark. i guess the CsvMeta is unnecessary, but we just need a way to deal with conflicting values